---
title: "16S rRNA analysis of BSF gut microbiota"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# First round of alpha and beta diversity statistical analysis
```
library(tidyverse)
library(vegan)
library(BiocManager)
library(phyloseq)
library(ANCOMBC)
library(DESeq2)


otu <- read.table(file = "feature-table.tsv", sep = "\t", header = T, row.names = 1, 
                  skip = 1, comment.char = "")
taxonomy <- read.table(file = "taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


metadata <- read.table(file = "metadata.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)  
TAX = tax_table(as.matrix(tax.clean))  
SAMPLE <- sample_data(metadata)  
TREE = read_tree("tree.nwk")  
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)  
```

# PCoA
```
dist = phyloseq::distance(ps, method="bray")
ordination = ordinate(ps, method="PCoA", distance=dist)
plot_ordination(ps, ordination, color="diet", shape = "line") + 
  theme_classic() +
  theme(strip.background = element_blank())
  ```

# PERMANOVA/ADONIS
```
metadata <- data.frame(sample_data(ps))
pairwise_adonis_results <- adonis2(dist~diet, method = "bray", data= data.frame( diet = sample_data(ps)$diet) )

pairwise_adonis_results
```

# PAIRWISE PERMANOVA
```
view(metadata)

cbn <- combn(x=unique(metadata$samples), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
  ps.subs <- subset_samples(ps, samples %in% cbn[,i])
  metadata_sub <- data.frame(sample_data(ps.subs))
  permanova_pairwise <- adonis2(phyloseq::distance(ps.subs, method = "bray") ~ samples, 
                                data = metadata_sub)
  p <- c(p, permanova_pairwise$aov.tab$`Pr(>F)`[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
p.table
```

# Alpha diversity
```
plot_richness(ps, x="diet", measures=c("Observed", "Shannon")) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))


rich = estimate_richness(ps, measures = c("Observed", "Shannon"))
wilcox.observed <- pairwise.wilcox.test(rich$Observed, 
                                        sample_data(ps)$expid, 
                                        p.adjust.method = "BH")
tab.observed <- wilcox.observed$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.observed
```
# Shannon
```
options(max.print=999999)

wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
                                       sample_data(ps)$expid, 
                                       p.adjust.method = "BH")
tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.shannon
```
# Shannon for diets
```
wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
                                       sample_data(ps)$group, 
                                       p.adjust.method = "BH")
tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.shannon
```

# Within line differences
```
ps.sub <- subset_samples(ps, line %in% c("WT"))

wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
                                       sample_data(ps.sub)$diet, 
                                       p.adjust.method = "BH")
tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.shannon
```

# Pairwise anosim for line
```
cbn <- combn(x=unique(metadata$line), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
  ps.subs <- subset_samples(ps, line %in% cbn[,i])
  metadata_sub <- data.frame(sample_data(ps.subs))
  permanova_pairwise <- anosim(phyloseq::distance(ps.subs, method="jaccard", binary = TRUE), 
                               metadata_sub$line)
  p <- c(p, permanova_pairwise$signif[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)


# Pairwise anosim for line and diet

cbn <- combn(x=unique(metadata$expid), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
  ps.subs2 <- subset_samples(ps, expid %in% cbn[,i])
  metadata_sub2 <- data.frame(sample_data(ps.subs2))
  permanova_pairwise2 <- anosim(phyloseq::distance(ps.subs2, method="jaccard", binary = TRUE), 
                                metadata_sub2$expid)
  p <- c(p, permanova_pairwise2$signif[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
```

# DEseq and ANCOMBC for differential analysis

**Using the same phyloseq object as above**
```
tse = mia::makeTreeSummarizedExperimentFromPhyloseq(ps)

tse <- transformCounts(tse, method = "relabundance")

tse_genus <- agglomerateByRank(tse, rank = "Genus")

ds2 <- DESeqDataSet(tse_genus, ~diet)

dds <- DESeq(ds2)

res<- results(dds)
df <- as.data.frame(res)
```

**Adds taxon column that includes names of taxa**
```df$taxon <- rownames(df)```

**Orders the rows of data frame in increasing order firstly based on column
"log2FoldChange" and secondly based on "padj" column**
  
```
df <- df %>% arrange(log2FoldChange, padj)

knitr::kable(head(df)) %>% 
  kableExtra::kable_styling("striped") %>% 
  kableExtra::scroll_box(width = "100%")
  ```

# ANCOMBC analysis to compare with DEseq

```
pseq_genus <- phyloseq::tax_glom(ps, taxrank = "Genus")

out = ancombc(
  phyloseq = pseq_genus, 
  formula = "diet", 
  p_adj_method = "fdr", 
  zero_cut = 0.90, # by default prevalence filter of 10% is applied
  lib_cut = 0, 
  group = "diet", 
  struc_zero = TRUE, 
  neg_lb = TRUE, 
  tol = 1e-5, 
  max_iter = 100, 
  conserve = TRUE, 
  alpha = 0.05, 
  global = TRUE
)
res <- out$res

knitr::kable(res$diff_abn) %>% kableExtra::kable_styling("striped") %>% 
  kableExtra::scroll_box(width = "100%")



print(paste0("DESeq2 p-values under 0.05: ", sum(df$padj<0.05, na.rm = TRUE), "/", length(df$padj)))
```

**DESeq2 p-values under 0.05: 19/100**

```
df <- df[ !stringr::str_detect(df$taxon, "Genus:uncultured"), ]
highest3 <- df[df$padj %in% sort(df$padj, decreasing = TRUE)[1:3], ]$taxon
highest3 <- assay(tse_genus, "clr")[highest3, ]
highest3 <- t(highest3)
highest3 <- data.frame(highest3, as.data.frame(colData(tse_genus)))
colnames(highest3)[1:3] <- lapply(colnames(highest3)[1:3], function(x){
  
  temp <- stringr::str_replace(x, "[.]", ": ")
  
  temp <- stringr::str_replace_all(temp, c("[.]" = " ", "_" = " "))
  
  temp <- stringr::str_wrap(temp, width = 25)
})

```
**Same but for taxa with lowest p-values**
```
lowest3 <- df[df$padj %in% sort(df$padj, decreasing = FALSE)[1:3], ]$taxon
lowest3 <-assay(tse_genus, "clr")[lowest3, ]
lowest3 <- t(lowest3)
lowest3 <- data.frame(lowest3, as.data.frame(colData(tse_genus)))
colnames(lowest3)[1:3] <- lapply(colnames(lowest3)[1:3], function(x){
 
  temp <- stringr::str_replace(x, "[.]", ": ")
  
  temp <- stringr::str_replace_all(temp, c("[.]" = " ", "_" = " "))
  
  temp <- stringr::str_wrap(temp, width = 25)
})
```
**Plot 1**
```
gridExtra::grid.arrange(
  
  ggplot(highest3, aes(x = diet, y = highest3[,1])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(highest3)[1]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title
```
  
**Plot 2**
```
ggplot(highest3, aes(x = diet, y = highest3[,2])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(highest3)[2]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title
```
**Plot 3**
```
  ggplot(highest3, aes(x = diet, y = highest3[,3])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(highest3)[3]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title
```
**Plot 4**
```
ggplot(lowest3, aes(x = diet, y = lowest3[,1])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(lowest3)[1]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title
```  
**Plot 5**
```
ggplot(lowest3, aes(x = diet, y = lowest3[,2])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(lowest3)[2]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title
```
**Plot 6**
```
ggplot(lowest3, aes(x = diet, y = lowest3[,3])) + 
    geom_boxplot() + 
    ylab("CLR abundances") + # y axis title
    ggtitle(names(lowest3)[3]) + # main title
    theme(title = element_text(size = 7),
          axis.text = element_text(size = 7),
          axis.title.x=element_blank()), # makes titles smaller, removes x axis title

ncol = 3, nrow = 2)
```

# DEseq
```
sample_data(ps)$diet <- as.factor(sample_data(ps)$diet) # factorize for DESeq2
ps.taxa <- tax_glom(ps, taxrank = 'Genus', NArm = FALSE)
```

**filter sparse features, with > 90% zeros**
```
ps.taxa.pse.sub <- prune_taxa(rowSums(otu_table(ps.taxa) == 0) < ncol(otu_table(ps.taxa)) * 0.9, ps.taxa)
ps_ds = phyloseq_to_deseq2(ps.taxa.pse.sub, ~ diet)
```
# use alternative estimator on a condition of "every gene contains a sample with a zero"
```
ds <- estimateSizeFactors(ps_ds, type="poscounts")
ds = DESeq(ds, test="Wald", fitType="parametric")
alpha = 0.05 
res = results(ds, alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
taxa_sig = rownames(res[1:10, ]) # select bottom 10 with lowest p.adj values
ps.taxa.rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)
ps.taxa.rel.sig <- prune_taxa(taxa_sig, ps.taxa.rel)
```

**plot**
```
matrix <- as.matrix(data.frame(otu_table(ps.taxa.rel.sig)))
rownames(matrix) <- as.character(tax_table(ps.taxa.rel.sig)[, "Genus"])
metadata_sub <- data.frame(sample_data(ps.taxa.rel.sig))

annotation_col = data.frame(
  Line = as.factor(metadata_sub$line), 
  `Diet` = as.factor(metadata_sub$diet), 
  check.names = FALSE
)
rownames(annotation_col) = rownames(metadata_sub)

annotation_row = data.frame(
  Phylum = as.factor(tax_table(ps.taxa.rel.sig)[, "Phylum"])
)
rownames(annotation_row) = rownames(matrix)

phylum_col = RColorBrewer::brewer.pal(length(levels(annotation_row$Phylum)), "Paired")
names(phylum_col) = levels(annotation_row$Phylum)
ann_colors = list(
  Line = c(`Line-WT` = "red", `Line-LD` = "blue"),
  `Diet` = c(CF = "purple", PKM = "yellow", D0="green",MLP="blue",MHP="purple",RIB="pink"),
  Phylum = phylum_col
)

ComplexHeatmap::pheatmap(matrix, scale= "row", 
                         annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = ann_colors)
```

  
# PCOA plots using the same phyloseq object

**PCOA: for each diet based on experimental days**

**CF**
```
ps.sub.CTRL <- subset_samples(ps, diet %in% c("D0", "CF"))

dist = phyloseq::distance(ps.sub.CTRL, method="bray")
ordination = ordinate(ps.sub.CTRL, method="PCoA", distance=dist)

p1<-plot_ordination(ps.sub.CTRL, ordination, color="experimental.day", shape = "line") + 
  theme_bw()+ geom_point(size=3)+ ggtitle("CF") + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())


print(p1)
```
**NUS**
```
ps.sub.NUS <- subset_samples(ps, diet %in% c("D0", "NUS"))

dist = phyloseq::distance(ps.sub.NUS, method="bray")
ordination = ordinate(ps.sub.NUS, method="PCoA", distance=dist)

p2<- plot_ordination(ps.sub.NUS, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("NUS") + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())


print(p2)
```
**MHP**
```
ps.sub.MHP <- subset_samples(ps, diet %in% c("D0", "MHP"))

dist = phyloseq::distance(ps.sub.MHP, method="bray")
ordination = ordinate(ps.sub.MHP, method="PCoA", distance=dist)

p3<- plot_ordination(ps.sub.MHP, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("MHP")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p3)
```
**MLP**
```
ps.sub.MLP <- subset_samples(ps, diet %in% c("D0", "MLP"))

dist = phyloseq::distance(ps.sub.MLP, method="bray")
ordination = ordinate(ps.sub.MLP, method="PCoA", distance=dist)

p4 <- plot_ordination(ps.sub.MLP, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("MLP")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p4)
```
**M**
```
ps.sub.M <- subset_samples(ps, diet %in% c("D0", "M"))

dist = phyloseq::distance(ps.sub.M, method="bray")
ordination = ordinate(ps.sub.M, method="PCoA", distance=dist)

p5 <- plot_ordination(ps.sub.M, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("M")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p5)
```
**P**
```
ps.sub.P <- subset_samples(ps, diet %in% c("D0", "P"))

dist = phyloseq::distance(ps.sub.P, method="bray")
ordination = ordinate(ps.sub.P, method="PCoA", distance=dist)

p6 <- plot_ordination(ps.sub.P, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("P")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p6)
```
**OKA**
```
ps.sub.OKA <- subset_samples(ps, diet %in% c("D0", "OKA"))

dist = phyloseq::distance(ps.sub.OKA, method="bray")
ordination = ordinate(ps.sub.OKA, method="PCoA", distance=dist)

p7 <- plot_ordination(ps.sub.OKA, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("OKA")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p7)
```
**PKM**
```
ps.sub.PKM <- subset_samples(ps, diet %in% c("D0", "PKM"))

dist = phyloseq::distance(ps.sub.PKM, method="bray")
ordination = ordinate(ps.sub.PKM, method="PCoA", distance=dist)

p8 <- plot_ordination(ps.sub.PKM, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("PKM")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p8)
```
**RIB**
```
ps.sub.RIB <- subset_samples(ps, diet %in% c("D0", "RIB"))

dist = phyloseq::distance(ps.sub.RIB, method="bray")
ordination = ordinate(ps.sub.RIB, method="PCoA", distance=dist)

p9 <- plot_ordination(ps.sub.RIB, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("RIB")  + stat_ellipse(aes(group = experimental.day), linetype = 2) +
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p9)
```
**SBM**
```
ps.sub.SBM <- subset_samples(ps, diet %in% c("D0", "SBM"))

dist = phyloseq::distance(ps.sub.SBM, method="bray")
ordination = ordinate(ps.sub.SBM, method="PCoA", distance=dist)

p10 <- plot_ordination(ps.sub.SBM, ordination, color="experimental.day", shape = "line") + geom_point(size=3) +
  theme_bw() + ggtitle("SBM")  + stat_ellipse(aes(group = experimental.day), linetype = 2) + 
  scale_color_manual(values=c("orange", "deeppink", "brown4"),limits = c("D0", "D5", "D10"))+ labs(shape="Line", colour="Experimental Day")+ scale_shape_manual(values=c(16, 1))+
  theme(strip.background = element_blank())

print(p10)



ggarrange(p1, p2, p8, p10, p5, p3, p9, p6, p7, p4, nrow=5, ncol=2, common.legend = TRUE, legend="bottom")
```


  
# Shannon plot [^1]
```
library(tidyverse)
library(qiime2R)
library(ggplot2)
library(ggpubr)
library(grid)


metadata<-read_q2metadata("metadata.tsv")
shannon<-read_qza("shannon_vector.qza")

view(metadata)

shannon<-shannon$data %>% rownames_to_column("SampleID")

view(shannon)

gplots::venn(list(metadata=metadata$SampleID, shannon=shannon$SampleID))

metadata1<- metadata %>% left_join(shannon)

view(metadata1)
head(metadata1)
```
**Plot for D0 and D5**
```
met2 <- metadata1[c(1:33,64:95),c(1:17)]
view(met2)

met2$diet_f = factor(met2$diet, levels=c('D0','CF','OKA','PKM','RIB','SBM', 'M', 'P', 'MHP', 'MLP', 'NUS'))
legend_title <- "Line"

plot1<-met2 %>%
  filter(!is.na(shannon)) %>%
  ggplot(aes(x= line, y=shannon, fill=line)) +
  stat_summary(geom="bar", fun.data=mean_se, color="black") +
  geom_jitter(shape=21, width=0.1, height=0) +
  coord_cartesian(ylim=c(0,7)) +
  facet_grid(~diet_f) +
  xlab("Diet") +
  ylab("Shannon Diversity") +
  theme_q2r() +
  scale_fill_manual(legend_title, values=c("black","white")) +
  theme(legend.position="right") + theme(axis.text.x=element_blank())+ theme(text=element_text(size = 15)) 
```
**Plot for D0 and D10**
```
met3 <- metadata1[c(1:3,34:66,96:125),c(1:17)]
view(met3)

met3$diet_f = factor(met3$diet, levels=c('D0','CF','OKA','PKM','RIB','SBM', 'M', 'P', 'MHP', 'MLP', 'NUS'))
legend_title <- "Line"

plot2<-met3 %>%
  filter(!is.na(shannon)) %>%
  ggplot(aes(x= line, y=shannon, fill=line)) +
  stat_summary(geom="bar", fun.data=mean_se, color="black") +
  geom_jitter(shape=21, width=0.1, height=0) +
  coord_cartesian(ylim=c(0,7)) +
  facet_grid(~diet_f) +
  xlab("Diet") +
  ylab("Shannon Diversity") +
  theme_q2r() +
  scale_fill_manual(legend_title, values=c("black","white")) + 
  theme(legend.position="right") + theme(axis.text.x=element_blank())+ theme(text=element_text(size = 15)) 



plot <- ggarrange(plot1 + rremove("ylab") + rremove("xlab"), plot2 + rremove("ylab") + rremove("xlab"), # remove axis labels from plots
                  ncol = 1, nrow = 2,
                  common.legend = TRUE, legend = "bottom")



annotate_figure(plot, left = textGrob("Shannon Diversity", rot = 90, vjust = 0.5, gp = gpar(cex = 1.3)),
                bottom = textGrob("Diet", gp = gpar(cex = 1.3)))
```


# Taxa abundance boxplot [^2]
```
library(tidyverse)
library(BiocManager)
library(vegan)
library(phyloseq)
library(microbiome)
library(RColorBrewer)
library(ggpubr)
library(dplyr)
library(microbiomeutilities)
library(grid)
library(gridExtra)

otu <- read.table(file = "feature-table.tsv", sep = "\t", header = T, row.names = 1, 
                  skip = 1, comment.char = "")
taxonomy <- read.table(file = "taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


metadata <- read.table(file = "metadata.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)  
TAX = tax_table(as.matrix(tax.clean))  
SAMPLE <- sample_data(metadata)  
TREE = read_tree("tree.nwk")  
ps <- phyloseq(OTU, TAX, SAMPLE,TREE) 
```
**Phylum prevalence summary**
```
glom <- tax_glom(ps, taxrank = 'Phylum')
percentages <- psmelt(glom)
View(percentages)
str(percentages)
write.table(as.data.frame(percentages), file = "percentage.xls", sep = "\t")

```
**Genus prevalence summary**
```
glom <- tax_glom(ps, taxrank = 'Genus')
percentages <- psmelt(glom)
View(percentages)
str(percentages)
write.table(as.data.frame(percentages), file = "percentagegenus.xls", sep = "\t")
```
**Differential abundance analysis**
```
sample_data(ps)$line <- as.factor(sample_data(ps)$line) # factorize for DESeq2
ps.taxa <- tax_glom(ps, taxrank = 'Genus', NArm = FALSE)

**DeSeq2**
ps.taxa.sub <- subset_samples(ps.taxa, type %in% c("WT", "LD"))
# filter sparse features, with > 90% zeros
ps.taxa.pse.sub <- prune_taxa(rowSums(otu_table(ps.taxa.sub) == 0) < ncol(otu_table(ps.taxa.sub)) * 0.9, ps.taxa.sub)
ps_ds = phyloseq_to_deseq2(ps.taxa.pse.sub, ~ line)
# use alternative estimator on a condition of "every gene contains a sample with a zero"
ds <- estimateSizeFactors(ps_ds, type="poscounts")
ds = DESeq(ds, test="Wald", fitType="parametric")
alpha = 0.05 
res = results(ds, alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
taxa_sig = rownames(res[1:20, ]) # select bottom 20 with lowest p.adj values
ps.taxa.rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)
ps.taxa.rel.sig <- prune_taxa(taxa_sig, ps.taxa.rel)

ps.taxa.rel.sig <- prune_samples(colnames(otu_table(ps.taxa.pse.sub)), ps.taxa.rel.sig)
ps.taxa.rel.sig
```

**Phylum boxplot**
```
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})


ps_phylum <- phyloseq::tax_glom(ps_rel_abund, "Phylum")
phyloseq::taxa_names(ps_phylum) <- phyloseq::tax_table(ps_phylum)[, "Phylum"]
#phyloseq::otu_table(ps_phylum)[1:5, 1:5]
ps_phylum <- microbiomeutilities::aggregate_top_taxa2(ps_phylum, "Phylum", top = 5)


p<-phyloseq::psmelt(ps_phylum) %>%
  ggplot(data = ., aes(x=diet, y = Abundance)) +
  geom_boxplot(outlier.shape  = NA) +
  geom_jitter(aes(shape = line)) + theme_bw()+ theme(legend.position="none")+
  scale_shape_manual(values = c(WT = 1,LD = 16)) +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~ OTU, scales = "free")

p + ggtitle("A")+ ylim(0,1.0) +scale_x_discrete(limits = c("D0", "CF", "MHP", "MLP","M","NUS","OKA","PKM","P","RIB","SBM"))

```
**Genus boxplot**
```
table(phyloseq::tax_table(ps)[, "Genus"])
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})
phyloseq::otu_table(ps)[1:5, 1:5]
phyloseq::otu_table(ps_rel_abund)[1:5, 1:5]
ps_genus <- phyloseq::tax_glom(ps_rel_abund, "Genus")
phyloseq::taxa_names(ps_genus) <- phyloseq::tax_table(ps_genus)[, "Genus"]
phyloseq::otu_table(ps_genus)[1:5, 1:5]

ps_genus <- microbiomeutilities::aggregate_top_taxa2(ps_genus, "Genus", top = 14)

variable_names <- list(
  "Dysgonomonas" ="Dysgonomonas",
  "Enterococcus_H_360604" = "Enterococcus",
  "Klebsiella_724518" = "Klebsiella",
  "Lacrimispora" ="Lacrimispora",
  "Lactiplantibacillus"="Lactiplantibacillus",
  "Levilactobacillus"="Levilactobacillus",
  "Morganella"="Morganella",
  "Other"="Other",
  "Paenibacillus_J_366884" = "Paenibacillus",
  "Proteus"="Proteus",
  "Providencia_A_732258"= "Providencia",
  "Scrofimicrobium"="Scrofimicrobium",
  "Sphingobacterium"="Sphingobacterium",
  "Unclassified Enterobacteriaceae_A" = "Unclassified Enterobacteriaceae",
  "Unclassified Enterococcaceae"="Unclassified Enterococcaceae")


variable_labeller <- function(variable,value){
  return(variable_names[value])
}


p2<-phyloseq::psmelt(ps_genus) %>%
  ggplot(data = ., aes(x = diet, y = Abundance)) +
  geom_boxplot(outlier.shape  = NA) +
  geom_jitter(aes(shape = line)) + theme_bw()+ 
  scale_shape_manual(values = c(WT = 1,LD = 16)) +
  labs(x = "", y = "Relative Abundance\n") + theme(legend.position="none")+
  facet_wrap(~ OTU, scales = "free",ncol = 3,labeller=variable_labeller)


p2 + ggtitle("B") +scale_x_discrete(limits = c("D0", "CF", "MHP", "MLP","M","NUS","OKA","PKM","P","RIB","SBM"))
```

# One-way ANOVA  
```
  library(dplyr)

data <- read.csv(file = "one way anova absolute relative abundance.csv", header = TRUE)

data.subset <- subset(data, substrate %in% c("Veg","Meat"))

View(data.subset)


anova_result <- aov(Proteobacteria ~ substrate, data=data.subset)
summary(anova_result)

anova_result2 <- aov(Actinobacteriota +Bacteroidota ~ substrate, data=data.subset)
summary(anova_result2)

data.subset2 <- subset(data, line %in% c("WT","LD"))
data.OKA <- subset(data.subset2, diet %in% "OKA")

View(data.OKA)

anova_result3 <- aov(Bacteroidota ~ line, data=data.OKA)
summary(anova_result3)
```
**Genus**
```
data2 <- read.csv(file = "genus one way anova absolute relative abundance.csv", header = TRUE)


data.age <- subset(data2, age %in% c("D5", "D10"))

View(data.age)
```
**one-way ANOVA**
```
anova_result_age1 <- aov(Paenibacillus_C ~ age, data=data.age)
summary(anova_result_age1)

anova_result_age2 <- aov(Lysinibacillus_304693 ~ age, data=data.age)
summary(anova_result_age2)

data.subset <- subset(data2, substrate %in% c("Veg","Meat"))

View(data.subset)
```
**one-way ANOVA**
```
anova_result <- aov(Scrofimicrobium + Sphingobacterium ~ substrate, data=data.subset)
summary(anova_result)

anova_result4 <- aov(Providencia_A_732258 ~ substrate, data=data.subset)
summary(anova_result4)

anova_result5 <- aov(Proteus ~ substrate, data=data.subset)
summary(anova_result5)

anova_result6 <- aov(Morganella ~ substrate, data=data.subset)
summary(anova_result6)


anova_result7 <- aov(Klebsiella_724518 ~ substrate, data=data.subset)
summary(anova_result7)

anova_result8 <- aov(Klebsiella_724518 ~ line, data=data2)
summary(anova_result8)

anova_result9 <- aov(Enterococcus_H_360604 + Morganella + Proteus + Providencia_A_732258 + Levilactobacillus + Scrofimicrobium  ~ line, data=data2)
summary(anova_result9)

anova_result10 <- aov(Lactiplantibacillus + Paenibacillus_J_366884 + Lacrimispora ~ line, data=data2)
summary(anova_result10)
```

# Beta diversity calculations
 ``` 
library(vegan)
library(ggplot2)
library(tidyverse)
library(RColorBrewer)
library(tidyverse)
library(BiocManager)
library(phyloseq)
library(ANCOMBC)
library(DESeq2)


otu <- read.table(file = "feature-table.tsv", sep = "\t", header = T, row.names = 1, 
                  skip = 1, comment.char = "")
taxonomy <- read.table(file = "taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


metadata <- read.table(file = "metadata2.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)  
TAX = tax_table(as.matrix(tax.clean))  
SAMPLE <- sample_data(metadata)  
TREE = read_tree("tree.nwk")  
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)  



relab_genera = transform_sample_counts(ps, function(x) x / sum(x) * 100)

abrel_bray <- phyloseq::distance(relab_genera, method = "bray")
abrel_bray <- as.matrix(abrel_bray)
head(abrel_bray)[,1:6]
```
**Save the matrix as a table**
```
write.table(abrel_bray, file = "Final_bray_curtis_result.txt", sep = "\t", col.names = NA)


df <- read.csv("betadiversity.csv", header = TRUE)

view(df)

averages <- df %>%
  group_by(Diet, Age) %>%
  summarise(
    Average_Between_Line = mean(Between.Line),
    Average_Within_Line = mean(Within.Line),
    SD_Between_Line = sd(Between.Line),
    SD_Within_Line = sd(Within.Line)
  )

merged_data <- averages %>%
  pivot_longer(
    cols = starts_with("Average"),
    names_to = "Line_Type",
    values_to = "Average_Value"
  ) %>%
  mutate(
    Line_Type = gsub("Average_", "", Line_Type),
    Legend = ifelse(Line_Type == "Between_Line", "Between Line", "Within Line")
  )


ggplot(merged_data, aes(x = Diet, y = Average_Value, shape = Line_Type, color = Age)) +
  geom_point(position = position_dodge(width = 0.5), size = 3) +
  geom_errorbar(
    aes(
      ymin = Average_Value - SD_Between_Line,
      ymax = Average_Value + SD_Between_Line
    ),
    position = position_dodge(width = 0.5),
    width = 0.2
  ) +
  labs(
    title = "Genetic and Environmental Effects on Bacterial Diversity",
    x = "Diet",
    y = "Average Value of Beta Diversity",
    color = "Larval Age",
    shape = "Beta Diversity"
  ) +
  scale_color_manual(values = c("D5" = "deeppink", "D10" = "brown4"),limits = c("D5", "D10")) +
  scale_shape_manual(values = c("Between_Line" = 16, "Within_Line" = 17)) +
  theme_minimal()
```

# ANOVA models for variables [^3]
  ```
library(mvabund)

meta <- read.csv(file = "L6clean3.csv", header = TRUE)

View(meta)

meta_spp <- mvabund(meta[, 7:270])

par(mar = c(2, 10, 2, 2)) # adjusts the margins
boxplot(meta[, 7:270], horizontal = TRUE, las = 2, main = "Abundance")


meanvar.plot(meta_spp)

plot(meta_spp ~ as.factor(meta$Diet), cex.axis = 0.8, cex = 0.8)

mod1 <- manyglm(meta_spp ~ meta$Diet, family = "poisson")
plot(mod1) #not evenly distributed. not good

mod2 <- manyglm(meta_spp ~ meta$Diet, family = "negative_binomial")
plot(mod2) #This residual plot is much better, there is now no discernible fan shape and we will use this model for all further analysis.

anova(mod2)

plot1<- anova(mod2, p.uni = "adjusted")
```

**test for multiple variables at the same time**
```
mod3 <- manyglm(meta_spp ~ meta$gut.mass, family = "negative_binomial")
plot(mod3)
anova(mod3)

mod4 <- manyglm(meta_spp ~ meta$biomass, family = "negative_binomial")
plot(mod4)
anova(mod4)

mod5 <- manyglm(meta_spp ~ meta$bioconversion, family = "negative_binomial")
anova(mod5)

mod6 <- manyglm(meta_spp ~ meta$endpoint, family = "negative_binomial")
anova(mod6)


mod7 <- manyglm(meta_spp ~ meta$Diet*meta$Line*meta$Substrate*meta$Larval.age, family = "negative_binomial")
anova(mod7)
```
# PERMANOVA

```
genus_matrix <- as.matrix(genus)

data_combined <- cbind(meta[, c("Diet", "Line", "Substrate", "Larval.age")], genus_matrix)
```
**Perform PERMANOVA**
```
permanova_result <- adonis2(genus_matrix ~ Diet + Line + Larval.age, data = data_combined, permutations = 999)

print(permanova_result)


  
library(tidyverse)
library(vegan)
library(phyloseq)
library(ANCOMBC)
library(DESeq2)
library(ComplexHeatmap)
library(qiime2R)
library(ggplot2)
library(labdsv)
library(DECIPHER)
library(ape)
library(plotly)
library(philr)
library(adespatial)
library(devtools)
library(MicrobeR)
library(microbiome)
library(microbiomeSeq)
library(ranacapa)
library(grid)
library(gridExtra)
library(knitr)
library(ggdendro)
library(ggpubr)
library(RColorBrewer)
library(microbiomeutilities)
library(VennDiagram)
library(RColorBrewer)
library(dplyr)
library(reshape2)
library(kableExtra)
library(mia)
```
**Create phyloseq object (ps) as above**


# Alpha diversity
```
plot_richness(ps, x="diet", measures=c("Observed", "Shannon")) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```
**Alpha diversity for diet type**
```
plot_richness(ps, x="type", measures=c("Observed", "Shannon")) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```
**Alpha diversity for bsf line**
```
plot_richness(ps, x="line", measures=c("Observed", "Shannon")) +
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```

**Pairwise test with Wilcoxon rank-sum test, corrected by FDR method. For the number of observed tags**
```
rich = estimate_richness(ps, measures = c("Observed", "Shannon"))

wilcox.observed <- pairwise.wilcox.test(rich$Observed, 
                                        sample_data(ps)$diet, 
                                        p.adjust.method = "BH")

tab.observed <- wilcox.observed$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()

view(tab.observed)
```
**Pairwise test with Wilcoxon rank-sum test, corrected by FDR method:Shannon diversity**
```
wilcox.shannon <- pairwise.wilcox.test(rich$Shannon, 
                                       sample_data(ps)$diet, 
                                       p.adjust.method = "BH")

tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()

view(tab.shannon)
```
# Beta diversity

**ANOSIM**
```
metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$diet)

metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$expid)

metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$bsf.line)

metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$experimental.day)

metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$type)

metadata <- data.frame(sample_data(ps))
anosim(dist, metadata$group)
```
**PAIRWISE anosim: diet**
```
cbn <- combn(x=unique(metadata$expid), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
  ps.subs <- subset_samples(ps, expid %in% cbn[,i])
  metadata_sub <- data.frame(sample_data(ps.subs))
  permanova_pairwise <- anosim(phyloseq::distance(ps.subs, method="bray", binary = TRUE), 
                               metadata_sub$diet)
  p <- c(p, permanova_pairwise$signif[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)

view(p.table)
```

  
# RDA analysis [^4],[^5],[^6]
 ``` 
library(phyloseq)
library(ggplot2)
library(microViz)
library(vegan)
library(microbiomeutilities)
library(microbiome)
library(tidyverse)
library(stringr)

otu <- read.table(file = "feature-table.tsv", sep = "\t", header = T, row.names = 1, 
                  skip = 1, comment.char = "")
taxonomy <- read.table(file = "taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


metadata <- read.table(file = "3RDAmeta.txt", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)  
TAX = tax_table(as.matrix(tax.clean))  
SAMPLE <- sample_data(metadata)  
TREE = read_tree("tree.nwk")  
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)


ps.diet <- subset_samples(ps, diet %in% c("CF", "OKA","PKM","MLP","RIB","P","NUS","SBM","M"))


renamer <- function(x) str_replace(x, pattern = "_", replacement = " ")

p<- ps.diet %>%
  tax_transform("hellinger", rank = "Genus") %>%
  ord_calc(
    constraints = c("fat", "protein", "carb","ash"),
    # method = "RDA", 
    scale_cc = FALSE
  ) %>%
  ord_plot(
    colour = "diet", size = 3, alpha = 0.5, shape = "experimental.day", auto_caption = NA,
    plot_taxa = 1:10, taxon_renamer = renamer, tax_vec_length = 2.5, tax_lab_length = 2.6,
    tax_lab_style = tax_lab_style(
      type = "text", max_angle = 90, fontface = "bold.italic", size = 3),
    constraint_lab_style = constraint_lab_style(
      alpha = 0.8, size = 3.5))+ 
  coord_fixed(ratio = 1, clip = "off", xlim = c(-1, 1.4))


p + stat_ellipse(aes(linetype = Substrate, colour=diet)) + scale_shape_manual(name= "Larval age", values = c("D5"=0,"D10"=15))+ scale_color_manual(name= "Diet", values = c("CF"="darkgoldenrod1","OKA"="red","P"="deeppink","PKM"="darkorange","RIB"="darkgreen","SBM"="chartreuse","MLP"="cyan3","M"="darkblue","NUS"="darkorchid"))

print(p)
```
# Pearson correlation [^7]
```
library(microeco)
library(magrittr)

view(metadata)
```
**create a microtable object**
```
dataset <- microtable$new(sample_table = metadata, otu_table = otu, tax_table = tax.clean, phylo_tree = TREE)
dataset

t1 <- trans_env$new(dataset = dataset, env_cols = 14:17)
```
**ANOVA**
```
t1$cal_diff(group = "diet", method = "anova")
head(t1$res_diff)


t1$cal_diff(group = "experimental.day", method = "anova")
head(t1$res_diff)


t1$cal_diff(group = "line", method = "anova")
head(t1$res_diff)
```
**boxplot of ANOVA (Diet)**

```
t1$cal_diff(method = "anova", group = "diet")

tmp <- list()
for(i in colnames(t1$data_env)){
  tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 5, xtext_size = 10) + theme(plot.margin = unit(c(0.1, 0, 0, 1), "cm"))
}
plot(gridExtra::arrangeGrob(grobs = tmp, ncol = 2))
```
**boxplot of ANOVA (experimental day)**

```
t1$cal_diff(method = "anova", group = "experimental.day")
# place all the plots into a list
tmp <- list()
for(i in colnames(t1$data_env)){
  tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 5, xtext_size = 12) + theme(plot.margin = unit(c(0.1, 0, 0, 1), "cm"))
}
plot(gridExtra::arrangeGrob(grobs = tmp, ncol = 2))
```
**boxplot of ANOVA (line)**

```
t1$cal_diff(method = "anova", group = "line")
# place all the plots into a list
tmp <- list()
for(i in colnames(t1$data_env)){
  tmp[[i]] <- t1$plot_diff(measure = i, add_sig_text_size = 5, xtext_size = 12) + theme(plot.margin = unit(c(0.1, 0, 0, 1), "cm"))
}
plot(gridExtra::arrangeGrob(grobs = tmp, ncol = 2))


library(GGally)

t1$cal_autocor()

t1$cal_autocor(group = "experimental.day")

t1$cal_autocor(group = "line")

t1$cal_autocor(group = "diet")
```

**'p_adjust_type = "Env"' means p adjustment is performed for each environmental variable separately**

```
t1$cal_cor(use_data = "Genus", p_adjust_method = "fdr", p_adjust_type = "Env")

t1$plot_cor()
```
**filter genera that do not have at least one (***)**

```
plot <- t1$plot_cor(filter_feature = c("", "*", "**"))

plot <- plot + theme(text = element_text(size = 14))

print(plot)
```

**RDA analysis (db-RDA and RDA)**

**use bray-curtis distance for dbRDA**

```
t1$cal_ordination(method = "dbRDA", use_measure = "bray")

t1$trans_ordination()
t1$plot_ordination(plot_color = "diet",color_values = RColorBrewer::brewer.pal(11, "Set3"))
```
**the main results of RDA are related with the projection and angles between arrows**

```
t1$trans_ordination(adjust_arrow_length = TRUE, max_perc_env = 1.5)

t1$plot_ordination(plot_color = "diet",color_values = RColorBrewer::brewer.pal(11, "Set3"))

t1$cal_ordination_anova()
t1$cal_ordination_envfit()

view(t1$res_ordination_axis)
view(t1$res_ordination_terms)

t1$cal_mantel(use_measure = "bray")

head(t1$res_mantel)
```

# KEGG Metabolic gene prediction [^8], [^9], [^10]

```
library(heatmaply)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(hrbrthemes)
library(viridis)
library(plotly)
library(ggpubr)
```

**Heatmap level 1 and 2**

```
level1<- read.delim("Tablelevel1.txt", header = TRUE)
view(level1)

colnames(level1) <- gsub("\\.", " ", colnames(level1))
```

**Matrix format**
```
mat1 <- level1
rownames(mat1) <- mat1[,1]
mat1 <- mat1 %>% dplyr::select(-Diet)
mat1 <- as.matrix(mat1)

heatmaply(mat1, 
          dendrogram = "none",
          xlab = "", ylab = "", 
          main = "",
          margins = c(60,100,40,20),
          scale="none",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 9, fontsize_col = 9,
          labCol = colnames(mat1),
          labRow = rownames(mat1),
          heatmap_layers = theme(axis.line=element_blank()))



level2<- read.delim("Tablelevel2.txt", header = TRUE)

view(level2)

colnames(level2) <- gsub("\\.", " ", colnames(level2))


mat2 <- level2
rownames(mat2) <- mat2[,1]
mat2 <- mat2 %>% dplyr::select(-Diet)
mat2 <- as.matrix(mat2)


heatmaply(mat2, 
          dendrogram = "none",
          xlab = "", ylab = "", 
          main = "",
          margins = c(60,100,40,20),
          scale="none",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 9, fontsize_col = 9,
          labCol = colnames(mat2),
          labRow = rownames(mat2),
          heatmap_layers = theme(axis.line=element_blank()))

```

# MANOVA to observe significance of Line and/or diet on metabolic functions
```
dat<- read.delim("newlevel1.txt", header = TRUE)
view(dat)


res.man <- manova(cbind(Cellular.Processes,
                        Environmental.Information.Processing,
                        Genetic.Information.Processing,
                        Human.Diseases,Metabolism, Organismal.Systems) ~ Diet+Line+Diet*Line, data = dat)
summary(res.man)

summary.aov(res.man)


dat2<- read.delim("newlevel2.txt", header = TRUE)
view(dat2)


res.man2 <- manova(cbind(Amino.Acid.Metabolism,
                         Biosynthesis.of.Other.Secondary.Metabolites,
                         Carbohydrate.Metabolism,
                         Glycan.Biosynthesis.and.Metabolism,
                         Infectious.Diseases, Metabolism,
                         Lipid.Metabolism,Metabolism.of.Cofactors.and.Vitamins,
                         Metabolism.of.Other.Amino.Acids,
                         Metabolism.of.Terpenoids.and.Polyketides,
                         Nucleotide.Metabolism,
                         Xenobiotics.Biodegradation.and.Metabolism,
                         Energy.Metabolism) ~ Diet+Line+Diet*Line, data = dat2)
summary(res.man2)

summary.aov(res.man2)
```
# Heatmap of significant metabolic functions based on diet and line
```

data<- read.delim("level1_sig.txt", header = TRUE)

colnames(data) <- gsub("\\.", " ", colnames(data))
view(data)

mat <- data
rownames(mat) <- mat[,1]
mat <- mat %>% dplyr::select(-Diet)
mat <- as.matrix(mat)

library(heatmaply)
heatmaply(mat, 
          dendrogram = "none",
          xlab = "", ylab = "", 
          main = "",
          margins = c(60,100,40,20),
          scale="column",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 9, fontsize_col = 9,
          labCol = colnames(mat),
          labRow = rownames(mat),
          heatmap_layers = theme(axis.line=element_blank()))



data<- read.delim("llevel2_sig.txt", header = TRUE)

colnames(data) <- gsub("\\.", " ", colnames(data))
view(data)

mat <- data
rownames(mat) <- mat[,1]
mat <- mat %>% dplyr::select(-Diet)
mat <- as.matrix(mat)

library(heatmaply)
heatmaply(mat, 
          dendrogram = "none",
          xlab = "", ylab = "", 
          main = "",
          margins = c(60,100,40,20),
          scale="column",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 9, fontsize_col = 9,
          labCol = colnames(mat),
          labRow = rownames(mat),
          heatmap_layers = theme(axis.line=element_blank()))
```
# Heatmap of significant metabolic functions based on diet only (for main figure)
```

data<- read.delim("diet_level1_sig.txt", header = TRUE)

colnames(data) <- gsub("\\.", " ", colnames(data))
view(data)

mat <- data
rownames(mat) <- mat[,1]
mat <- mat %>% dplyr::select(-Diet)
mat <- as.matrix(mat)

library(heatmaply)
heatmaply(mat, 
          dendrogram = "none",
          xlab = "", ylab = "", 
          main = "",
          margins = c(60,100,40,20),
          scale="column",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 12, fontsize_col = 12,
          labCol = colnames(mat),
          labRow = rownames(mat),
          heatmap_layers = theme(axis.line=element_blank()))



data<- read.delim("diet_llevel2_sig.txt", header = TRUE)

colnames(data) <- gsub("\\.", " ", colnames(data))
view(data)

mat <- data
rownames(mat) <- mat[,1]
mat <- mat %>% dplyr::select(-Diet)
mat <- as.matrix(mat)

library(heatmaply)
heatmaply(mat, 
          dendrogram = "none",
          xlab = "", ylab = "",
          main = "",
          margins = c(60,100,40,20),
          scale="column",
          grid_color = "white",
          grid_width = 0.00001,
          titleX = FALSE,
          hide_colorbar = FALSE,
          branches_lwd = 0.1,
          label_names = c("Diet", "Feature:", "Value"),
          fontsize_row = 12, fontsize_col = 12,
          labCol = colnames(mat),
          labRow = rownames(mat),
          heatmap_layers = theme(axis.line=element_blank()))
```


# Core microbiota analysis [^11]
```
library(eulerr)
library(heatmaply)

otu <- read.table(file = "feature-table.tsv", sep = "\t", header = T, row.names = 1, 
                  skip = 1, comment.char = "")
taxonomy <- read.table(file = "taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "d__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


metadata <- read.table(file = "metadata.tsv", sep = "\t", header = T, row.names = 1)
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)  
TAX = tax_table(as.matrix(tax.clean))  
SAMPLE <- sample_data(metadata)  
TREE = read_tree("tree.nwk")  
ps <- phyloseq(OTU, TAX, SAMPLE,TREE)  
```

**detection is set to 0 to detect even the rarest microbe**

**line 50%**
  
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c() # an empty object to store information

for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub, 
                         detection = 0, 
                         prevalence = 0.50)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m 
}

print(list_core)

mycols <- c(LD="dark grey", WT="white") 

line50 <-  c(LD = 16, WT = 7, "LD&WT" = 20)
fit1 <- euler(line50, shape = "ellipse")
plot1<-plot(fit1, quantities = TRUE,
            fill = mycols,
            labels = list(font = 4), main = "50%")
```


**line 60%**
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c()

for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub, 
                         detection = 0, 
                         prevalence = 0.60)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}

print(list_core)

mycols <- c(LD="dark grey", WT="white")
mycols <- c(LD="dark grey", WT="white") 
#plot(venn(list_core),fills = mycols)

line60 <-  c(LD = 15, WT = 6, "LD&WT" = 16)
fit2 <- euler(line60, shape = "ellipse")

plot2 <- plot(fit2, quantities = TRUE,
              fill = mycols,
              labels = list(font = 4), main = "60%")
```
**line 70%**
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c()


for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub, 
                         detection = 0, 
                         prevalence = 0.70)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}

print(list_core)

mycols <- c(LD="dark grey", WT="white") 
#plot(venn(list_core),fills = mycols)

line70 <-  c(LD = 16, WT = 4, "LD&WT" = 12)
fit3 <- euler(line70, shape = "ellipse")
plot3 <- plot(fit3 , quantities = TRUE,
              fill = mycols,
              labels = list(font = 4), main = "70%")
```
**line 80%**
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c() # an empty object to store information


for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub, 
                         detection = 0, 
                         prevalence = 0.80)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}

print(list_core)

mycols <- c(LD="dark grey", WT="white") 
#plot(venn(list_core),fills = mycols)

line80 <-  c(LD = 14, WT = 3, "LD&WT" = 6)
fit4 <- euler(line80, shape = "ellipse")
plot4 <- plot(fit4 , quantities = TRUE,
              fill = mycols,
              labels = list(font = 4), main = "80%")

```
**line 90%**
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c() # an empty object to store information


for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub,  
                         detection = 0, 
                         prevalence = 0.90)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}

print(list_core)

mycols <- c(LD="dark grey", WT="white") 
#plot(venn(list_core),fills = mycols)

line90 <-  c(LD = 10, WT = 0, "LD&WT" = 3)
fit5 <- euler(line90, shape = "ellipse")
plot5 <- plot(fit5 , quantities = TRUE,
              fill = mycols,
              labels = list(font = 4), main = "90%")
```

**line 99.9%**
```
table(meta(ps)$line, useNA = "always")

pseq.rel <- microbiome::transform(ps, "compositional")

lines <- unique(as.character(meta(pseq.rel)$line))
print(lines)

list_core <- c() # an empty object to store information


for (n in lines){ # for each variable n in line
  #print(paste0("Identifying Core Taxa for ", n))
  
  ps.sub <- subset_samples(pseq.rel, line == n)
  
  core_m <- core_members(ps.sub,  
                         detection = 0,  
                         prevalence = 0.999)
  print(paste0("No. of core taxa in ", n, " : ", length(core_m)))
  list_core[[n]] <- core_m # add to a list core taxa for each group.
  #print(list_core)
}

print(list_core)

mycols <- c(LD="dark grey", WT="white") 
#plot(venn(list_core),fills = mycols)

line99.9 <-  c(LD = 2, WT = 0, "LD&WT" = 1)
fit6 <- euler(line99.9, shape = "ellipse")
plot6 <- plot(fit6 , quantities = TRUE,
              fill = mycols,
              labels = list(font = 4), main = "99.9%")


plot <- ggarrange(plot1, plot2, plot3, plot4, plot5, plot6, nrow=2, ncol=3)

annotate_figure(plot, top = text_grob("Comparison of Prevalence Thresholds", 
                                      color = "black", face = "bold", size = 20))
```

# Tutorials
[^1]: https://forum.qiime2.org/t/tutorial-integrating-qiime2-and-r-for-data-visualization-and-analysis-using-qiime2r/4121
[^2]: https://www.nicholas-ollberding.com/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/
[^3]: https://environmentalcomputing.net/statistics/mvabund/
[^4]: https://fawda123.github.io/ggord/reference/ggord.html
[^5]: https://david-barnett.github.io/microViz/articles/web-only/ordination.html#rda
[^6]: https://david-barnett.github.io/microViz/reference/ord_plot.html
[^7]: https://chiliubio.github.io/microeco_tutorial/explainable-class.html#trans_env-class (Section 3 and 7)
[^8]: https://www.scribbr.com/statistics/anova-in-r/
[^9]: http://www.sthda.com/english/wiki/manova-test-in-r-multivariate-analysis-of-variance
[^10]: https://www.data-to-viz.com/graph/heatmap.html
[^11]: https://microbiome.github.io/tutorials/core_venn.html, https://microbiome.github.io/tutorials/Core.html